### PR TITLE
issue 57: allow customers to register and to log in

### PR DESCRIPTION
**Pull request by jkonan100**
@michowski, @claudist  Could you please review?

In this item:
= we add a command to register new customers
= add a command to "login"
= small refactoring

Kindest Regards,
Jules

### CODE REVIEW

File: src/main/java/newbank/server/Customer.java [hunk 1]
```diff
@@ -40,4 +40,15 @@ public String getUsername() {
   public String getPassword() {
     return password;
   }
+  
+  @Override
+  public String toString() {
+	  StringBuilder sb = new StringBuilder();
+	  sb.append(this.getClass().getName() + "[");
```
Comment (michowski) [inline src/main/java/newbank/server/Customer.java:47]: No newlines in this output? Does this look readable in the current form?

File: src/main/java/newbank/server/NewBank.java [hunk 2]
```diff
@@ -42,10 +45,24 @@ private void addTestData() {
    * @param username The customer's username
    * @param password The customer's password
    */
-  public void addCustomer(final String username, final String password) {
+  public void addCustomer(final String username, final String password)
+      throws DuplicateCustomerException, PasswordInvalidException {
+    if (customers.containsKey(username)) {
+      throw new DuplicateCustomerException();
```
Comment (michowski) [inline src/main/java/newbank/server/NewBank.java:51]: Perfect!

Comment (michowski) [reply src/main/java/newbank/server/NewBank.java:51]: Perhaps, to be consistent, it should be `CustomerDuplicateException`. Again - I do realize it doesn't sounds the most natural.

File: src/main/java/newbank/server/NewBank.java [hunk 3]
```diff
@@ -42,10 +45,24 @@ private void addTestData() {
...
+
+    validatePassword(password);
+
     customers.put(username, new Customer(username, password));
   }
 
+  // Simple algorithm to check that the password meets the security requirements
+  private void validatePassword(final String password) throws PasswordInvalidException {
+    if (password.isEmpty()) {
+      throw new PasswordInvalidException();
```
Comment (michowski) [inline src/main/java/newbank/server/NewBank.java:62]: Possibly could be `CustomerPasswordInvalidException` - unless we are considering a general notion of a password. But again - I do realize I'm bikeshedding a lil bit :))

Not a blocker either way!

File: src/main/java/newbank/server/NewBank.java [hunk 4]
```diff
@@ -66,56 +83,67 @@ public synchronized CustomerID checkLogInDetails(final String username, final St
...
+   * @param customerID The customer identifier
+   * @return account information
+   */
+  public synchronized String showAccountsFor(final CustomerID customerID) {
+    Customer customer = customers.get(customerID.getKey());
+
+    return showAccountsFor(customer);
+  }
+
+  private String showAccountsFor(final Customer customer) {
```
Comment (michowski) [inline src/main/java/newbank/server/NewBank.java:98]: Is there any value of this overload? Not sure if we need both methods.

File: src/main/java/newbank/server/NewBankClientHandler.java [hunk 5]
```diff
@@ -6,46 +6,128 @@
...
+      username = tokens[1];
+    }
+    if (tokens.length >= 3) {
+      password = tokens[2];
+    }
+
+    try {
+      bank.addCustomer(username, password);
+
+      out.println("SUCCESS: Account created successfully.");
```
Comment (michowski) [inline src/main/java/newbank/server/NewBankClientHandler.java]: Account in this context is a little ambiguous. `Customer account` perhaps? Or: `Registered successfully.`?

File: src/main/java/newbank/server/NewBankClientHandler.java [hunk 6]
```diff
@@ -6,46 +6,128 @@
...
+    } catch (DuplicateCustomerException ex) {
+      out.println("FAIL: " + String.format("FAIL: Customer name [%s] already exists.", username));
```
Comment (michowski) [inline src/main/java/newbank/server/NewBankClientHandler.java]: A redundant fail string: `FAIL: FAIL:`

File: src/main/java/newbank/server/NewBankClientHandler.java [hunk 7]
```diff
@@ -6,46 +6,128 @@
...
+      out.println("FAIL: Specified password does not meet the security requirements.");
+    }
+  }
+
+  private void handleShowAccounts() throws RequestNotAllowedException {
+    checkLoggedIn();
+    out.println(bank.showAccountsFor(customer));
+  }
+
+  private void checkLoggedIn() throws RequestNotAllowedException {
```
Comment (michowski) [inline src/main/java/newbank/server/NewBankClientHandler.java:74]: I'm thinking if there is an alternative approach we could use... but I guess that is only possible once every command is a separate class. For now, I cannot come up with anything better.

File: src/main/java/newbank/server/NewBankClientHandler.java [hunk 8]
```diff
@@ -6,46 +6,128 @@
...
+    }
+  }
+
+  private boolean processRequest(final String request) {
+    final String[] tokens = request.split("\\s+");
+
+    assert (tokens.length > 0);
+
+    try {
+      switch (tokens[0].toUpperCase()) {
```
Comment (michowski) [inline src/main/java/newbank/server/NewBankClientHandler.java:86]: It's surely an UX improvement, but also a little bit of a global change to the protocol. Perhaps we shall document it.

File: src/main/java/newbank/server/NewBankClientHandler.java [hunk 9]
```diff
@@ -6,46 +6,128 @@
...
+  }
+
+  private void checkLoggedIn() throws RequestNotAllowedException {
+    if (customer == null) {
+      throw new RequestNotAllowedException();
+    }
+  }
+
+  private boolean processRequest(final String request) {
+    final String[] tokens = request.split("\\s+");
```
Comment (michowski) [inline src/main/java/newbank/server/NewBankClientHandler.java:81]: Similarly as here: https://github.com/Group14-Newbank/newbank/pull/19/files#r602767465

File: src/main/java/newbank/server/exceptions/CustomerMaxAccountsException.java [hunk 10]
```diff
@@ -1,5 +1,5 @@
 package newbank.server.exceptions;
 
 public class CustomerMaxAccountsException extends Exception {
-
+  private static final long serialVersionUID = 1234564L;
```
Comment (michowski) [inline src/main/java/newbank/server/exceptions/CustomerMaxAccountsException.java:4]: I don't think there is a point in adding them (regardless of the warnings) unless we are:
- utilizing them in serialization
- they are actually unique

**Review by michowski — COMMENTED** [review]
Hi Jules,

great work as always. I'm just slightly concerned about the amount of changes this PR introduces. Especially that different changes seem to be touching different aspects of the codebase and the product.

In general, this could cause a lot of (git) conflicts when 3 developers work in parallel.

Left a few comments, some of them aren't blockers.

Best,
Arek

File: src/main/java/newbank/server/exceptions/CustomerMaxAccountsException.java (hunk 10 above)
Comment (juliano60) [reply src/main/java/newbank/server/exceptions/CustomerMaxAccountsException.java:4]: The IDE adds them in automatically if you allow it too. It also get rids of compiler warnings.

Kindest Regards,
Jules

File: src/main/java/newbank/server/NewBankClientHandler.java (hunk 7 above)
Comment (juliano60) [reply src/main/java/newbank/server/NewBankClientHandler.java:74]: I am moving towards a separate class for each command. In fact that's the object of my next PR.

Kindest Regards,
Jules

File: src/main/java/newbank/server/NewBankClientHandler.java (hunk 6 above)
Comment (juliano60) [reply src/main/java/newbank/server/NewBankClientHandler.java]: Well spotted!

File: src/main/java/newbank/server/NewBankClientHandler.java (hunk 5 above)
Comment (juliano60) [reply src/main/java/newbank/server/NewBankClientHandler.java]: Done.

File: src/main/java/newbank/server/NewBank.java (hunk 4 above)
Comment (juliano60) [reply src/main/java/newbank/server/NewBank.java:98]: Agreed that the overloads need re-visiting at some point. 

Kindest Regards,
Jules

File: src/main/java/newbank/server/NewBank.java (hunk 3 above)
Comment (juliano60) [reply src/main/java/newbank/server/NewBank.java:62]: Done.

File: src/main/java/newbank/server/NewBank.java (hunk 2 above)
Comment (juliano60) [reply src/main/java/newbank/server/NewBank.java:51]: It does not read well I don't think.

File: src/main/java/newbank/server/Customer.java (hunk 1 above)
Comment (juliano60) [reply src/main/java/newbank/server/Customer.java:47]: The to_string does not add newlines. The client will add newlines if needed.

**Comment by juliano60** [comment]
Thanks Arek!

**Review by michowski — DISMISSED** [review]
Let's goooo! :rocket:

**Comment by michowski** [comment]
A few conflicts to resolve though :)))

**Comment by juliano60** [comment]
Caught with main and resolved conflicts.

Kindest Regards,
Jules

**Pull request merged**